rfcs/0006-transaction.md

```diff
@@ -0,0 +1,299 @@
+
+Given an execution history of transactions like:
+
+```
+txn1: [seq:97] <--- ---- ---> [seq:99]
+txn2:     [seq:98]  <--- ---- ---- ---- ---- ---- ---> [seq:105]
+txn3:                         [seq:102] <--- ---> [seq:104]
+txn4:                    [seq:101] <--- ---- ---- ---- ---> [seq:106]
+```
+
+Let's say the current transaction is `txn4`, it should check conflicts with `txn2` and `txn3` during the commit, because `txn2` and `txn3` are committed after `txn4` started. During the conflict checking, `txn4`'s Read & Write keys should not conflict with the Write keys of `txn2` and `txn3`.
+
+It does not need to check conflicts with `txn1`, because `txn1` is committed before `txn4` started. This could tell us when to GC the entry in the `recent_committed_txns` deque: when ALL the running transactions' sequence number are bigger than the committed sequence number of the entry, then the entry could be GCed.
+
+Please note that `read_keys` are represented as a set of `Range<Bytes>`. This allows us to efficiently track ranges of keys that were read during the transaction, and ensure Serializable Snapshot Isolation (SSI) on range conflicts. For example, if transaction A reads a range `["key01", "key10"]` and transaction B writes to `"key05"` which falls within that range, we can detect this conflict and abort one of the transactions to maintain serializability.
+
+This requires we track the original write keys in the `TransactionState`, instead of the key fingerprints like Badger does. We could reuse the buffers of the original keys in WriteBatch to reference the write keys with slices of `Bytes`.
+
+This approach can be adjusted to support SI as well by only tracking the keys that are being written to. When user set the isolation level to SI, the `read_keys` could always be set as empty. The GC mechanism to release inactive transactions is still the same as the SSI.
+
+### How to Test
+
+Testing transaction implementations requires verifying both correctness and isolation guarantees.
+
+To test the transaction feature, we could write some test cases that cover the following aspects:
+
+1. Basic transaction operations.
+2. Isolation guarantees for both SI and SSI modes.
+3. Edge cases and concurrency control.
+
+We could consider to port the test cases from [RocksDB](https://github.com/facebook/rocksdb/blob/3495c94761629404f5476f732e2130134cfcc51b/utilities/transactions/optimistic_transaction_test.cc), [Badger](https://github.com/dgraph-io/badger/blob/36c461a435c53a8a81e7377c2b026b24d37eee0c/txn_test.go) to have a comprehensive coverage of the above aspects.
+
+## Roadmap
+
+1. Implement the WriteBatch API (related with https://github.com/slatedb/slatedb/pull/264).
+2. Add sequence to each key in the storage, and implement the Snapshot API.
+3. Add the conflict checking in the transaction, implement the SI transaction.
```

Comment:
Reading the previous sections, specifically the rocksdb section, it looks like SI would be implemented by reading the keys in write set, and comparing the sequence number. 
**Conflict checking** section on the other hand says below
"""
This approach can be adjusted to support SI as well by only tracking the keys that are being written to. 
"""

Given we are going to implement the global oracle for SSI, implementing global oracle alone is a reasonable option, or at least one that we could start with. Is that the intention? Can this decision be added explicitly?

Reply:
make sense, updated in 4f1b8db, ptal